### PR TITLE
Optimize the data loader for datasets that fit in memory or on a local disk

### PR DESCRIPTION
**Pull request by cwognum**
## Changelogs

- Added `Dataset.load_in_memory()`, which optimizes how a Zarr archive is stored _if_ it fits in main memory.
- Added a new tutorial to highlight the effect of this function and to give general optimization tips.
- Changed from `df.loc[row, col]` to `df.at[row, col]` to optimize Pandas indexing. 
- Use memory mapping to load data locally.
- Use a default chunk size of 1 in the `SDFConverter` to have a chunk per data point. 

---

_Checklist:_

- [ ] ~_Was this PR discussed in an issue? It is recommended to first discuss a new feature into a GitHub issue before opening a PR._~
- [X] _Add tests to cover the fixed bug(s) or the newly introduced feature(s) (if appropriate)._
- [X] _Update the API documentation if a new function is added, or an existing one is deleted._
- [X] _Write concise and explanatory changelogs above._
- [X] _If possible, assign one of the following labels to the PR: `feature`, `fix` or `test` (or ask a maintainer to do it for you)._

---

The changes in this PR were the results of some iterative profiling and optimization that was documented [here](https://docs.google.com/document/d/1Ed1rEkPja0OpRujocJbxHej0jep_3Erya6IelsgS7pw/edit?usp=sharing) and [here](https://docs.google.com/document/d/1gP-usLiLG7hqg-M0zE0n1KHFl7o6Z-1fIhsOok5NLjw/edit?usp=sharing).

### CODE REVIEW

**Review by lmtroper — APPROVED** [review]
Looks good to me! Changes make sense as outlined in the doc.

**Comment by cwognum** [comment]
Two things I want to think about a bit longer here: 
- [ ] Whether we want to make the assumption that the `Dataset.table` index is the default one (e.g. what about competitions). 
- [ ] Whether we want to overwrite `Dataset._zarr_root`. What if people create a dataset, optimize it and then save it? Is that a realistic scenario?

File: polaris/dataset/zarr/_memmap.py [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+import zarr
+
+
+class MemoryMappedDirectoryStore(zarr.DirectoryStore):
+    """
+    A Zarr Store to open chunks as memory-mapped files.
+    See https://github.com/zarr-developers/zarr-python/issues/1245
+
+    Memory mapping leverages low-level OS functionality to reduce the time it takes
+    to read the content of a file by directly mapping RAM memory to physical memory.
```
Comment (jstlaurent) [inline polaris/dataset/zarr/_memmap.py:12]: **Suggestion**
As above.
```suggestion
    to read the content of a file by directly mapping to memory.
```

**Review by jstlaurent — APPROVED** [review]
Starting to think we should write this in Rust and wrap it in PyO3. 😉

File: polaris/dataset/_dataset.py [hunk 2]
```diff
@@ -200,36 +206,61 @@ def client(self):
...
+        data loading, a user can choose to load the data into memory as a numpy array
+
+        Note: General purpose dataloader.
+            The goal with Polaris is to provide general purpose datasets that serve as good
+            options for a _wide variety of use cases_. This also implies you should be able to
+            optimize things further for a specific use case if needed.
+        """
+        if self._zarr_data is not None:
+            return self._zarr_data
+        return self.zarr_root
```
[the remark this answers is not in the document]
Comment (cwognum) [reply polaris/dataset/_dataset.py:223]: Yes! What is throwing you of?

File: polaris/dataset/_dataset.py [hunk 3]
```diff
@@ -254,7 +285,27 @@ def columns(self) -> list:
         """Return all columns for the dataset"""
         return self.table.columns.tolist()
 
-    def get_data(self, row: int, col: str, adapters: Optional[List[Adapter]] = None) -> np.ndarray:
+    def load_to_memory(self):
+        """Pre-load the entire dataset into RAM memory.
```
[the remark this answers is not in the document]
Comment (cwognum) [reply polaris/dataset/_dataset.py:289]: Good to know! I thought about this today and wasn't sure why people prefer "memory". It makes sense now! 👍

**Comment by cwognum** [comment]
> Starting to think we should write this in Rust and wrap it in PyO3. 😉

That seems to be the national tendency for most Python packages these days. I would have a lot to learn, but that sounds like fun!

File: polaris/dataset/_dataset.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (cwognum) [reply polaris/dataset/_dataset.py:223]: Discussed in person. We're good here!

**Pull request merged**